bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttBrokerConnection.java

```diff
@@ -49,81 +51,76 @@
-    private MqttAsyncClient client;
-    private boolean isConnecting = false;
-
-    private final List<MqttConnectionObserver> connectionObservers = new CopyOnWriteArrayList<>();
-    private final Map<String, List<MqttMessageSubscriber>> consumers = new HashMap<>();
-    // This should be removed by 2018 and before ESH 1.0
-    @Deprecated
-    private final List<MqttMessageProducer> producers = new CopyOnWriteArrayList<MqttMessageProducer>();
+    protected @Nullable MqttAsyncClient client;
+    protected boolean isConnecting = false;
+    protected final List<MqttConnectionObserver> connectionObservers = new CopyOnWriteArrayList<>();
+    protected final Map<String, List<MqttMessageSubscriber>> consumers = new HashMap<>();
 
     /**
      * A private object to implement the MqttCallback interface.
      * We don't want the MqttBrokerConnection to implement this directly.
-     *
+     * Developer hint: Unfortunately MqttCallback does not care about annotations for Null-correctness, we use local
+     * variables to compensate.
      */
     private class ClientCallbacks implements MqttCallback {
         @Override
-        public synchronized void connectionLost(Throwable t) {
-            if (t instanceof MqttException) {
-                MqttException e = (MqttException) t;
-                logger.info("MQTT connection to '{}' was lost: {} : ReasonCode {} : Cause : {}", getName(),
-                        e.getMessage(), e.getReasonCode(),
-                        (e.getCause() == null ? "Unknown" : e.getCause().getMessage()));
-            } else {
-                logger.info("MQTT connection to '{}' was lost: {}", getName(), t.getMessage());
+        public synchronized void connectionLost(@Nullable Throwable exception) {
+            if (exception instanceof MqttException) {
+                MqttException e = (MqttException) exception;
+                logger.info("MQTT connection to '{}' was lost: {} : ReasonCode {} : Cause : {}", host, e.getMessage(),
+                        e.getReasonCode(), (e.getCause() == null ? "Unknown" : e.getCause().getMessage()));
+            } else if (exception != null) {
+                logger.info("MQTT connection to '{}' was lost: {}", host, exception.getMessage());
             }
 
-            for (final MqttConnectionObserver connectionObserver : connectionObservers) {
-                connectionObserver.connectionStateChanged(MqttConnectionState.DISCONNECTED, t);
+            connectionObservers.forEach(o -> o.connectionStateChanged(MqttConnectionState.DISCONNECTED, exception));
+            if (reconnectStrategy != null) {
+                reconnectStrategy.lostConnection();
             }
-
-            reconnectStrategy.lostConnection();
         }
 
         @Override
-        public void deliveryComplete(IMqttDeliveryToken token) {
+        public void deliveryComplete(@Nullable IMqttDeliveryToken token_) {
+            IMqttDeliveryToken token = (@NonNull IMqttDeliveryToken) token_;
```

Comment:
this looks like an accident waiting to happen :-) Please simply null-check here.